presto-verifier/src/main/java/com/facebook/presto/verifier/rewrite/QueryRewriter.java

```diff
@@ -95,19 +95,32 @@ public class QueryRewriter
     private final PrestoAction prestoAction;
     private final Map<ClusterType, QualifiedName> prefixes;
     private final Map<ClusterType, List<Property>> tableProperties;
+    private final Optional<FunctionCallRewriter> functionCallRewriter;
 
     public QueryRewriter(
             SqlParser sqlParser,
             TypeManager typeManager,
             PrestoAction prestoAction,
             Map<ClusterType, QualifiedName> tablePrefixes,
             Map<ClusterType, List<Property>> tableProperties)
+    {
+        this(sqlParser, typeManager, prestoAction, tablePrefixes, tableProperties, Optional.empty());
+    }
+
+    public QueryRewriter(
+            SqlParser sqlParser,
+            TypeManager typeManager,
+            PrestoAction prestoAction,
+            Map<ClusterType, QualifiedName> tablePrefixes,
+            Map<ClusterType, List<Property>> tableProperties,
+            Optional<String> nonDeterministicFunctionSubstitutes)
     {
         this.sqlParser = requireNonNull(sqlParser, "sqlParser is null");
         this.typeManager = requireNonNull(typeManager, "typeManager is null");
         this.prestoAction = requireNonNull(prestoAction, "prestoAction is null");
         this.prefixes = ImmutableMap.copyOf(tablePrefixes);
         this.tableProperties = ImmutableMap.copyOf(tableProperties);
+        this.functionCallRewriter = requireNonNull(nonDeterministicFunctionSubstitutes, "nonDeterministicFunctionSubstitutes is null").map(FunctionCallRewriter::getInstance);
```

Comment:
I don't understand this:
1. nonDeterministicFunctionSubstitutes is Optional<String>, meaning it can be none.
2. We use requireNonNull, meaning is cannot be 'none'?

Reply:
For Java,
1) Optional<String> nonDeterministicFunctionSubstitutes = Optional.empty();
2) Optional<String> nonDeterministicFunctionSubstitutes = null;

are different. 2) cannot pass requireNonNull() while 1) can.

Separately, this line calls Optional.map(), and if FunctionCallRewriter.getInstance(nonDeterministicFunctionSubstitutes.get()) returns null, Optional.map() returns Optional.empty().